README.txt
= riakrest

http://github.com/wcpr/riakrest

== DESCRIPTION:

RiakRest provides structured, RESTful interaction with a Riak document
store.

== RIAK:

Riak[http://riak.basho.com] is an open-source project developed and maintained
by the fine folks at Basho[http://www.basho.com]. It combines a decentralized
key-value store, a flexible map/reduce engine, and a friendly HTTP/JSON query
interface to provide a database ideally suited for Web applications. RiakRest
interacts with the HTTP/JSON query interface, which is called Jiak.

== FEATURES/PROBLEMS:


== SYNOPSIS:

RiakRest provides structured, RESTful interaction with the HTTP/JSON interface
of a Riak[http://riak.basho.com] document data store. RiakRest provides two
levels of interaction: Core Client and Resource. Core Client works at the Jiak
level and exposes Jiak internals. JiakResource is an abstraction built on top
of the Core Client that gives a true RESTful feel.

== REQUIREMENTS:

RestClient is used for REST server interaction.<br/>
JSON is used for data exchange.<br/>
Riak[http://riak.basho.com] provides the HTTP/JSON Jiak interface and data store.

== INSTALL:

sudo gem install riakrest

== EXAMPLE
<code>
  require 'riakrest'
  include RiakRest
  
  PersonData = JiakDataHash.create(:name,:age)
  PersonData.keygen :name
  
  class Person
    include JiakResource
    server      'http://localhost:8002/jiak'
    group       'people'
    data_class  PersonData
    auto_post   true
    auto_update true
  end
  
  remy = Person.new(:name => 'remy', :age => 10)  #            (auto-post)
  remy.name                                       # => "remy"  (auto-update)

  puts Person.get('remy').name                    # => "remy"  (from Jiak server)
  puts Person.get('remy').age                     # => 10      (from Jiak server)
  
  remy.age = 11                                   #            (auto-update)
  puts Person.get('remy').age                     # => 11      (from Jiak server)
  
  callie = Person.new(:name => 'Callie', :age => 13)
  remy.link(callie,'sister')
  
  sisters = remy.query(Person,'sister')
  sisters[0].eql?(callie)                         # => true
  
  remy.delete
  callie.delete
</code>

Go forth and Riak!

== LICENSE:

(The MIT License)

Copyright (c) 2009 Paul Rogers, DingoSky LLC

Permission is hereby granted, free of charge, to any person obtaining
a copy of this software and associated documentation files (the
'Software'), to deal in the Software without restriction, including
without limitation the rights to use, copy, modify, merge, publish,
distribute, sublicense, and/or sell copies of the Software, and to
permit persons to whom the Software is furnished to do so, subject to
the following conditions:

The above copyright notice and this permission notice shall be
included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED 'AS IS', WITHOUT WARRANTY OF ANY KIND,
EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.
IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY
CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,
TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


History.txt
=== 0.1.7 2009-12-17
  - Added write data conversion.

=== 0.1.6 2009-12-17
  - Add method for specifying read data conversion in JiakData and JiakResouce.
  - Add testing and examples for data conversion.

=== 0.1.5 2009-12-11
  - Added POV processing.
  - Refactored jattr_* to attr_*.
  - Added opts to query, which forced steps arg to be an explicit array.
  - Changes to examples for web site.

=== 0.1.2 2009-11-29
  - Add wide open schemas.
  - Add wild card schema arrays.

=== 0.1.1 2009-11-28
  - Fixed several 1.8.7 issues.
  - Added helper for examples and put server URI setting in helpers.
  - Removed POV terminology. Riak doesn't support per-request masks (yet?).
  - Added some gem dependency declarations.
  - Minor refactoring.

=== 0.1.0 2009-11-16
  - Renamed data/JiakDataHash to core/JiakDataFields and reworked.
  - Move JiakDataHash functionality into JiakData as default implementation
    - Like Resource, JiakData has jattr_* to manage data inside out.
  - Resource encapsulates its data.
    - jattr_* methods add data fields and implicitly maintain schema.
    - Removed class methods for schema manipulation.
  - Incrementally add fields data.
  - Incrementally add fields to schemas.

=== 0.0.4 2009-11-10
  - Fixed bug when running ruby 1.8
  - Changed client_store option for returning object to :return => :object
  - Added check to eql? and == that both objects are either local or not.
  - Refactored local? method to jiak_object.

=== 0.0.3 2009-11-03
  - Switched to jeweler for managing gem
  - Allow passing an array for specifying schema fields/masks
  - Added convenience accessors to rsrc.jiak
  - Allow setting proxy on core client
  - More examples futzing.

=== 0.0.2 2009-10-30
  - Fixed corner case for auto_post false / auto_update true.
  - Explicitly reserve jiak field in user data classes.
  - Changed walk to query. Kept walk as an alias.
  - DRY'd some code
  - Examples


=== 0.0.1 2009-10-20
  - Initial release to github, anyway. git log shows the prior activity.


lib/riakrest.rb
# See README.rdoc for the RiakRest license.
begin
  require 'json'
rescue LoadError
  puts "\nRiakRest requires json for REST JSON messaging."
  puts "  Install with:  gem install json"
  puts
  exit
end

begin
  require 'restclient'
rescue LoadError
  puts "\nRiakRest requires the restclient gem for making REST calls."
  puts "  Install with:  gem install rest-client"
  exit
end

require 'uri'

dirname = File.dirname(__FILE__)
$:.unshift(dirname) unless
  $:.include?(dirname) || 
  $:.include?(File.expand_path(dirname))

# RiakRest provides structured, RESTful interaction with a Riak document data
# store. In Riak parlance, the HTTP/JSON interface to a Riak cluster is called
# Jiak. RiakRest provides two levels of Jiak interaction: Core Client and
# Resource. Core Client works directly with the Jiak level, whereas Resource is
# a resource-oriented abstraction on top of Core Client.
#
# ===Example
#  require 'riakrest'
#  include RiakRest
#
#  class People
#    include JiakResource
#    server 'http://localhost:8002/jiak'
#    attr_accessor :name, :age
#    auto_manage
#  end
#
#  remy = People.new(:name => 'Remy',:age => 10)       # (auto-post)
#  remy.age = 11                                       # (auto-update)
#
#  callie = People.new(:name => 'Callie', :age => 13)
#  remy.link(callie,'sister')
#
#  sisters = remy.query([People,'sister'])
#  sisters[0].eql?(callie)                             # => true
#
#  remy.delete
#  callie.delete
#
# ===Core Client
# See JiakClient for Core Client example.
#
# Go forth and Riak!
module RiakRest
  version_file = File.join(File.dirname(__FILE__),"..","VERSION")
  VERSION = IO.read(version_file).chomp

  # Convenience method for checking validity of method options. If any of the
  # options in opt are not in valid, raise the exception listing the invalid
  # options in the message.
  def check_opts(opts,valid,exception)              # :nodoc:
    unless(opts.empty?)
      err = opts.inject({}) do |h,(k,v)|
        h[k] = v  unless(valid.include?(k))
        h
      end
      unless err.empty?
        raise exception, "unrecognized options: #{err.keys}"
      end
    end
    opts
  end
end

require 'riakrest/core/exceptions'
require 'riakrest/core/jiak_bucket'
require 'riakrest/core/jiak_client'
require 'riakrest/core/jiak_data'
require 'riakrest/core/jiak_data_fields'
require 'riakrest/core/jiak_link'
require 'riakrest/core/jiak_object'
require 'riakrest/core/jiak_schema'
require 'riakrest/core/query_link'

require 'riakrest/resource/jiak_resource'
require 'riakrest/resource/jiak_resource_pov'

# Extend Array with convenience methods for comparing array contents.
class Array   # :nodoc:
  # Compare arrays for same elements regardless of order.
  def same_elements?(arr)
    raise ArgumentError unless arr.is_a?(Array)
    (size == arr.size) && arr.reduce(true){|same,elem| same && include?(elem)}
  end

  # Compare arrays for same element.to_s values regardless of order.
  def same_fields?(arr)
    raise ArgumentError unless arr.is_a?(Array)
    (size == arr.size) && map{|f| f.to_s}.same_elements?(arr.map{|f| f.to_s})
  end
end
